renku/core/models/provenance/datasets.py

```diff
@@ -377,13 +443,30 @@ def add(self, dataset):
 
     def get(self, identifier):
         """Return a dataset by its original identifier."""
-        datasets = [d for d in self._datasets if d.identifier == identifier]
-        assert len(datasets) <= 1, f"Found more than one with identifier `{identifier}`."
-        return datasets[0] if datasets else None
+        datasets = (d for d in self._datasets if d.identifier == identifier)
+        dataset = next(datasets, None)
+        assert next(datasets, None) is None, f"Found more than one dataset with identifier `{identifier}`."
```

Comment:
Should replace this assert with explicit check and raise?

Reply:
We've added these checks as safety-nets here and in other places to make sure that they never happen. I'll keep the format as it is to be in par with other places. End users should not see this (or the repo is really broken).